test/cpp/api/moduledict.cpp

```diff
@@ -0,0 +1,342 @@
+         const auto& rhs) {
+        return lhs.value().get() == rhs.value().get();
+      }));
+}
+
+TEST_F(ModuleDictTest, IsCloneable) {
+  torch::OrderedDict<std::string, std::shared_ptr<Module>> ordereddict = {
+    {"linear", Linear(2, 3).ptr()},
+    {"relu", Functional(torch::relu).ptr()},
+    {"batch", BatchNorm1d(3).ptr()},
+  };
+  ModuleDict dict(ordereddict);
+  ModuleDict clone = std::dynamic_pointer_cast<ModuleDictImpl>(dict->clone());
+  ASSERT_EQ(dict->size(), clone->size());
+
+  for (auto it = dict->begin(), it_c = clone->begin(); it != dict->end(); ++it, ++it_c) {
+    // The key should be same
+    ASSERT_EQ(it->key(), it_c->key());
+    // The modules should be the same kind (type).
+    ASSERT_EQ(it->value()->name(), it_c->value()->name());
+    // But not pointer-equal (distinct objects).
+    ASSERT_NE(it->value(), it_c->value());
+  }
+
+  // Verify that the clone is deep, i.e. parameters of modules are cloned too.
+  torch::NoGradGuard no_grad;
+
+  auto params1 = dict->named_parameters();
+  auto params2 = clone->named_parameters();
+  ASSERT_EQ(params1.size(), params2.size());
+  for (auto& param : params1) {
+    ASSERT_FALSE(pointer_equal(param.value(), params2[param.key()]));
+    ASSERT_EQ(param->device(), params2[param.key()].device());
+    ASSERT_TRUE(param->allclose(params2[param.key()]));
+    param->add_(2);
+  }
+  for (auto& param : params1) {
+    ASSERT_FALSE(param->allclose(params2[param.key()]));
+  }
+}
+
+TEST_F(ModuleDictTest, RegistersElementsAsSubmodules) {
+  torch::OrderedDict<std::string, std::shared_ptr<Module>> ordereddict1 = {
+    {"linear", Linear(10, 3).ptr()},
+    {"conv", Conv2d(1, 2, 3).ptr()},
+    {"test", Dropout(0.5).ptr()},
+  };
+  ModuleDict dict(ordereddict1);
+
+  auto modules = dict->children();
+  ASSERT_TRUE(modules[0]->as<Linear>());
+  ASSERT_TRUE(modules[1]->as<Conv2d>());
+  ASSERT_TRUE(modules[2]->as<Dropout>());
+
+  // Update Existing
+  torch::OrderedDict<std::string, std::shared_ptr<Module>> ordereddict2 = {
+    {"lstm", LSTM(4, 5).ptr()},
+    {"test", BatchNorm2d(5).ptr()}
+  };
+  dict->update(ordereddict2);
+
+  modules = dict->children();
+  ASSERT_TRUE(modules[0]->as<Linear>());
+  ASSERT_TRUE(modules[1]->as<Conv2d>());
+  // Keep Order
+  ASSERT_TRUE(modules[2]->as<BatchNorm2d>());
+  ASSERT_TRUE(modules[3]->as<LSTM>());
+}
+
+TEST_F(ModuleDictTest, CloneToDevice_CUDA) {
+  torch::OrderedDict<std::string, std::shared_ptr<Module>> ordereddict = {
+    {"linear", Linear(2, 3).ptr()},
+    {"relu", Functional(torch::relu).ptr()},
+    {"batch", BatchNorm1d(3).ptr()},
+  };
+  ModuleDict dict(ordereddict);
+  torch::Device device(torch::kCUDA, 0);
+  ModuleDict clone =
+      std::dynamic_pointer_cast<ModuleDictImpl>(dict->clone(device));
+  for (const auto& p : clone->parameters()) {
+    ASSERT_EQ(p.device(), device);
+  }
+  for (const auto& b : clone->buffers()) {
+    ASSERT_EQ(b.device(), device);
+  }
+}
+
+TEST_F(ModuleDictTest, IsCloneable_CUDA) {
```

Comment:
It seems to me that IsCloneable and IsCloneable_CUDA are pretty much the same except the device.
Can we do something like this so we don't need to maintain 2 pieces of identical code. Sry, I should mention this earlier.
Here is an example:
https://github.com/pytorch/pytorch/blob/e70c50bf59f3ca0d885a716043e2f31aea98be30/test/cpp/api/transformer.cpp#L152

Reply:
Yep, will do. Thanks for the advice!!